SPEC.md

```diff
@@ -0,0 +1,239 @@
+}
+```
+
+##### Chain
+
+The `chain` is stored as an `annotation` on the layer, in the same descriptor.
+The `annotation` key is `dev.cosignproject.cosign/chain`.
+
+Example `chain`:
+
+```
+"annotations": {
+      "dev.sigstore.cosign/chain": "-----BEGIN CERTIFICATE-----\nMIIB+DCCAX6gAwIBAgITNVkDZoCiofPDsy7dfm6geLbuhzAKBggqhkjOPQQDAzAq\nMRUwEwYDVQQKEwxzaWdzdG9yZS5kZXYxETAPBgNVBAMTCHNpZ3N0b3JlMB4XDTIx\nMDMwNzAzMjAyOVoXDTMxMDIyMzAzMjAyOVowKjEVMBMGA1UEChMMc2lnc3RvcmUu\nZGV2MREwDwYDVQQDEwhzaWdzdG9yZTB2MBAGByqGSM49AgEGBSuBBAAiA2IABLSy\nA7Ii5k+pNO8ZEWY0ylemWDowOkNa3kL+GZE5Z5GWehL9/A9bRNA3RbrsZ5i0Jcas\ntaRL7Sp5fp/jD5dxqc/UdTVnlvS16an+2Yfswe/QuLolRUCrcOE2+2iA5+tzd6Nm\nMGQwDgYDVR0PAQH/BAQDAgEGMBIGA1UdEwEB/wQIMAYBAf8CAQEwHQYDVR0OBBYE\nFMjFHQBBmiQpMlEk6w2uSu1KBtPsMB8GA1UdIwQYMBaAFMjFHQBBmiQpMlEk6w2u\nSu1KBtPsMAoGCCqGSM49BAMDA2gAMGUCMH8liWJfMui6vXXBhjDgY4MwslmN/TJx\nVe/83WrFomwmNf056y1X48F9c4m3a3ozXAIxAKjRay5/aj/jsKKGIkmQatjI8uup\nHr/+CxFvaJWmpYqNkLDGRU+9orzh5hI2RrcuaQ==\n-----END CERTIFICATE-----"
+}
+```
+
+## Payloads
+
+Implementations MUST support at least the following payload types:
+  
+  * Simple Signing
+
+### Simple Signing
+
+The Simple Signing payload format is [specified here](https://github.com/containers/image/blob/a5061e5a5f00333ea3a92e7103effd11c6e2f51d/docs/containers-signature.5.md#json-data-format)
+
+The following additional semantics are applied:
+
+* The `mediaType` used to identify this payload format is: `application/vnd.dev.cosign.simplesigning.v1+json`.
+* The `critical.type` value used to identify `cosign` signatures is: `cosign container signature`.
+* The `critical.identity.docker-reference` field is not supported.
+* Optional user-specified claims may be included in the `Optional` section.
+
+## Signature Schemes
+
+Implementations must support at least the following schemes:
+
+* ECDSA-P256 with the SHA256 Hash Algorithm
+
+No information about the signature scheme is included in the object.
+Clients must determine the hash algorithm and signature scheme out-of-band during the verification process.
+This is an intentional decision to reduce the risk of [algorithm-confusion attacks](https://news.ycombinator.com/item?id=24346317).
```

Comment:
I don't agree with this decision...

The cited issue is `alg=none`... which is a real JOSE issue... but that does not make JOSE by itself bad, just many, many implementations that blindly rely on it.

Above you say "ECDSA-P256 with the SHA256 Hash Algorithm" is required to be supported... why not say the signature formal will be Detached JWS with Unencoded payload and one of the following alg: "ES256, EdDsa", etc...

There are a lot of different signature standards, the only thing worse is no way to tell which one is used, when someone hands you a signed payload and no context.

consider the more popular signature suite specs, you might base this spec on:

https://tools.ietf.org/html/rfc4880#section-5.2.3

https://tools.ietf.org/html/rfc7515#section-4.1.1

^ notorious...

> The JWS Signature value is not
>  valid if the "alg" value does not represent a supported algorithm or
>  if there is not a key for use with that algorithm associated with the
>  party that digitally signed or MACed the content.

somehow ^ from this sentence, we got to "just turn off security with the `alg:none`"...

The commenters on the thread are correct.... if you are building a security library, don't give your callers a foot gun.... 

IMO, the solution is: 

- throw if algorithm is not defined
- throw if algorithm is not from allow list 
- update the allow list regularly based on security best practices ( https://safecurves.cr.yp.to/, etc)

This way you can know what software was signed with RSA 512 / MD5.... and not need to negotiate with someone out of band to learn that you should not trust that package.

We should assume that software will be published with crypto suites that are later broken, and make it easy to warn users.

Reply:
Thanks for the feedback! I think there are two parts to it, and we actually agree on the important parts. This spec was written to describe how things are today, not necessarily how they should be. I want to fix some of it before merging.

For the first part - we only support ecdsa-256 with sha256 right now. We should support more, especially for verification. I'm still leaning towards this implementation only supporting one type of key for signing, but we should be able to verify signatures of other types.

I'm not sure about this part though:

> There are a lot of different signature standards, the only thing worse is no way to tell which one is used, when someone hands you a signed payload and no context.

The signature is definitely meaningless if all you have is the signed payload and signature. You need a public key to verify against. I think what I'm trying to say is that we should include the information on the signing algorithm used with the public key, not the payload. 

We have a few components we need overall for the signature to be meaningful:

- Payload bytes
- Signature bytes
- Hash algorithm (sha256, etc.)
- Signature algorithm (ecdsa-p256, etc.)
- Public Key Bytes

In the current implementation, we have the following breakdown:

- Payload bytes (stored in registry, next to target)
- Signature bytes (stored in registry, next to target)
- Hash algorithm (sha256, etc.) (hardcoded to sha256)
- Signature algorithm (ecdsa-p256, etc.) (stored in PKIX public key, delivered out of band)
- Public Key Bytes (stored in PKIX public key, delivered out of band)

I agree the JOSE spec isn't bad, but I don't think it actually buys us anything. We don't need anything to be URL safe, and we don't need to concatenate everything into one token. It also would require us to duplicate the algorithm info (outside of the hash function), which is my real concern. If you have a public key you want to verify things against, you already know the algorithm.

So I guess the main question is - do we need to support multiple hash algorithms? If we do, let's figure out a place to include that info.

Reply:
While its not common, there are scenarios where having the public key is not enough,  for example: secp256k1 with ECDSA or Schnorr, or bls12381 with blake2b or shake256... Or Ed25519 to Curve25519 conversions, etc...

Most signatures schemes assume a hash algorithm, but it's also pretty common to hash with something different before signing.

My recommendation would be provide a data structure that allows you to specify everything needed to verify a signature without guessing.

Building on an existing signature suite and serialization scheme obviously helps make the software easier to implement in other languages with off the shelf tooling, in an interoperable way.

I would recommend building on JOSE, COSE, or PGP, and not inventing a new signature format.

But if you are inventing a new signature suite, it should cover the algorithmic agility, and assume more than one signature, hash and public key representation.

I think test vectors and a registry would be essential for this, something like:

```
0 ES256 (ECDSA with P256 and SHA256) on sha256(raw binary)
1 ES384 (ECDSA with P384 and SHA256) on sha256(raw binary)
```
^ saddly these will not produce stable test vectors due to non deterministic signatures.

You may also want to cite FIPS / NIST directly, since you can't rely on IANA JOSE  / COSE registries:

- https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-4.pdf
- https://www.iana.org/assignments/jose/jose.xhtml
- https://www.iana.org/assignments/cose/cose.xhtml

Reply:
To throw another signature spec into the mix: https://github.com/secure-systems-lab/signing-spec/blob/master/protocol.md

I think that one is closest philosophically to what I want to do here. It has support for "hinting" at the key id and algorithm alongside the payload, but it's very clear that this information can't be trusted and users are expected to obtain this information out of band.

Reply:
The only real difference between the current cosign implementation and the SSL spec is whether or not the payload type is "protected". Cosign uses the mediaType element to store the payload type (simple signing), and this media type is not protected by a signature.

If we used the SSL spec, we would instead set a media type corresponding to the SSL spec. Then the signed component would be:

` PAE(UTF8(SimpleSigning),$JSON)`

I'm not sure I understand the attack/threat model on putting this payload type into the signature.

Reply:
There are political considerations to promoting PAE, PGP, or really anything but JOSE / COSE... eventually the big guys will start attacking you since they are kinda locked to JOSE / COSE... 

I'd recommend sticking to JSON / JOSE... we should avoid innovation at the signature suite layer, and should focus on the input to the signature and the anchoring of the signature... I might go as far as saying the tool should support BOTH PGP and JOSE formatted signatures and keys.

meet developers where they are, with the formats they are familiar with.

Reply:
@dlorenc here is a really young spec, from the IoT space... https://www.ietf.org/archive/id/draft-devault-bare-01.txt

Maybe its worth considering as well.

Reply:
I’d argue fairly hard against designing a signature crypto system, or choosing an immature one. It’s not trivial, and it’s a huge distraction from container signing.

Find an industry standard that allows the container signing implementation to call a function with (root of trust in some widely-accepted form, signature blob, maybe (but probably not) an untrusted blob format string), and get a simple “signature valid/invalid” outcome (along with the signed payload if signature is valid).

That doesn’t _entirely_ simplify the discussion because choosing that industry standard still requires avoiding the incompetently-designed ones, and making a bit of a bet on its future viability, but it should at least getting the container signing implementation as far as possible out of the business of handling raw private keys, digest computations and parsing absolutely untrusted byte strings and https://golang.org/pkg/crypto/subtle/ .

With the MIME types in descriptors and the like it’s ultimately not _that big_ a deal to choose a wrong industry standard.

Reply:
+1 to that. This spec really deals with how the fields are stored in a container registry rather than any actual crypto details.

Reply:
Note: I am an author of the [SSL Signing Spec](https://github.com/secure-systems-lab/signing-spec), and I just filed #259 related to this topic.

> The only thing I don't really like about that spec is the usage of PAE.

@dlorenc Could you please explain your reservations?

> I'm not sure I understand the attack/threat model on putting this payload type into the signature.

If you have a signing oracle that uses the same key for multiple message types, then an attacker could potentially get the oracle to sign a message as one type and then have a verifier interpret it as another. Here's a very detailed example showing how to do that in a CI/CD scenario:
https://colab.research.google.com/github/secure-systems-lab/signing-spec/blob/master/hypothetical_signature_attack.ipynb#scrollTo=CIF11AmzXJuQ

Happy to explain more if that's not clear.

Reply:
My main concern with PAE is that there's very limited library, language, and tooling support for serialization and deserialization. I haven't actually been able to find any for Golang for example.

Reply:
PAE itself doesn't need a library. It's just a one-statement concatenation. You never deserialize it. I don't know go, but here it is in Python:

```python
def PAE(payloadType: str, payload: bytes) -> bytes:
  return b''.join([struct.pack('<Q', 2),
                   struct.pack('<Q', len(payloadType)),
                   payloadType.encode('utf-8'),
                   struct.pack('<Q', len(payload)),
                   payload])
```

Now, I agree that SSL Signing Spec needs a reference implementation, but I wanted to make sure there are no security concerns with PAE. It sounds like not.

Reply:
No security concerns, just interoperability. It's important we can do all these verifications in as much standard tooling as possible. This includes things like bash/openssl.

Reply:
Thanks for the feedback. I filed https://github.com/secure-systems-lab/signing-spec/issues/27 to consider a simpler PAE.